design/sync-v2.md

```diff
@@ -0,0 +1,216 @@
+* There are a variety of types here; we should artificially restrict the types.
+* The DTO should be able to be handled on both client and server, potentially mapping using Automapper to get the DatabaseGenerated field for UpdatedAt.
+
+This should also be handled quite easily with non-OData requests; for example, GraphQL, gRPC, and other transports should be easily provided.
+
+## Local Service
+
+On the local side, we should utilize common libraries, like Entity Framework Core, to implement the offline synchronization library.  This is split into two parts - 
+
+1. When the client application accesses the local database
+2. When the synchronization process makes changes to the local database.
+
+Ideally, the same DTO should be able to be used on both sides (remote service and offline database).  
+
+When a client application makes a change to the local database:
+
+1. The change is validated.
+2. The operation is written to the operations queue.
+3. The change is written to the local offline table.
+
+When the synchronization process makes a change to the local database:
+
+1. The change is validated.
+2. The operation is removed from the operations queue.
+3. The change is written to the local offline table.
+
+In both cases, the operations queue and offline table change needs to be written in a transaction so that it may be rolled back in its entirety. 
+
+In the new world:
+
+### Pull Operation
+
+1. Read delta-token from the offline table.
+2. Read new items since delta-token from remote service.
+3. Add new items to the offline table.
+4. Write delta-token to the offline table.
+
+### CRUD on offline table
+
+1. Add new entity to the offline table, writing an operation table entry as well (as transaction).
+    TODO: How to deal with Id field that is not globally unique in the offlien table.
+
+### Push Operation
+
+1. Read (block of) entities in the operations queue
+1. For each entity:
+    * Do operation on remote service.
+    * On success, store new entity + remote operation table entry as a transaction.
+    * On conflict, deal with conflicts as per current spec.
+    * On error, add error to aggregate exception, and leave operation table entry alone (abort queue).
+
+This relies on:
+
+* The operations queue is a part of the store - not a part of the sync context.
+* Errors and delta tokens are a part of the store as well.
+
+## Suggested API
+
+```csharp
+/// <remarks>
+/// Cancellation tokens are assumed.
+/// </remarks>
+public interface IOfflineTable<T>
+{
+    // Operations that are normal for client application to do:
+    IQueryable<T> AsQueryable();
+    ValueTask<Result<T>> GetAsync(object id);
+    ValueTask<IEnumerable<Result<T>>> AddRangeAsync(IEnumerable<T> entities);
+    ValueTask<IEnumerable<Result>> RemoveRangeAsync(IEnumerable<object> entityIds);
+    ValueTask<IEnumerable<Result<T>>> ReplaceRangeAsync(IEnumerable<T> entities);
+}
+
+internal interface ISynchronizableTable<T>
```

Comment:
Does this approach result in a memory instantiated table meta data object? With the `OperationsCount` per table the operation queue check during sync process could be short circuited in case of `OperationsCount == 0`!